Add gallery creator role setting and Restrict custom post type

Frenziecodes opened the pull request:
This commit updates the gallery registration process by implementing role-based capabilities. The register() function now registers the custom post type 'gallery' used by the FooGallery plugin with capabilities based on the selected gallery creator role. This ensures that only users with the appropriate role can perform actions such as creating, editing, and deleting galleries.

bradvin commented on includes/class-posttypes.php:
```diff
@@ -18,11 +18,20 @@ function __construct() {
 		function register() {
-			//allow extensions to override the gallery post type
-			$args = apply_filters( 'foogallery_gallery_posttype_register_args',
+			$foogallery_options   = get_option( 'foogallery' );
```
try using `foogallery_get_setting( 'gallery_creator_role ', array() )`. If the setting has not been set, then it should default to an empty array.

bradvin replied:
this var is not used, so this line can be removed

bradvin commented on includes/admin/class-settings.php:
```diff
@@ -118,6 +118,23 @@ function create_settings() {
+				'id'      => 'gallery_creator_role',
+				'title'   => __( 'Gallery Creator Role', 'foogallery' ),
+				'desc'    => __( 'Select the user role allowed to create galleries', 'foogallery' ),
+				'default' => 'administrator',
```
I do not think the default should be 'administrator'. This could potentially cause restrictions for customers who rely on it being available to non-admins by default

bradvin commented on includes/admin/class-settings.php:
```diff
@@ -118,6 +118,23 @@ function create_settings() {
 				'section' => __( 'Gallery Defaults', 'foogallery' )
 			);
 
+			$roles        = wp_roles()->roles;
```
consider using function `get_editable_roles` instead. I am not sure of the difference, but from the docs, it looks appropriate : https://developer.wordpress.org/reference/functions/get_editable_roles/

Frenziecodes replied:
Done some research and I think the main difference lies in how they retrieve and provide role information in WordPress. `get_editable_roles` retrieves only the roles that the current user can edit, based on their capabilities, while `wp_roles()->roles ` fetches all roles defined in the system, regardless of user permissions. In most cases, it is recommended to use get_editable_roles when you specifically need the editable roles for the current user.

bradvin commented on includes/class-posttypes.php:
```diff
@@ -34,23 +43,30 @@ function register() {
+					'show_ui'      => true,
+					'menu_icon'    => 'dashicons-format-gallery',
+					'supports'     => array( 'title', 'thumbnail' ),
+					'capabilities' => array(
```
you only want to add the 'capabilities' arg if the setting is set to something that is not the default

bradvin commented on includes/admin/class-settings.php:
```diff
@@ -118,6 +118,23 @@ function create_settings() {
+				'type'    => 'select',
+				'choices' => array_merge( array( 'administrator' => __( 'Administrator', 'foogallery' ) ), $role_choices ),
+				'tab'     => 'general',
+				'section' => __( 'Gallery Permissions', 'foogallery' ),
```
instead of adding a new section, rather use the already existing 'Admin' section

bradvin commented:
while testing this PR, I noticed that the Foogallery menu still appears because of the 'Albums' custom post type.
Please can you add a new setting for 'Album Creator Role' which can be added under the 'Albums' tab on the settings page. The default for this setting can be 'Inherit from gallery creator role' and then the options will be similar to the gallery creator role. If the setting is set to the default, then it will act the same as the gallery and what that role is set to.
You can then apply similar logic within `\extensions\albums\class-posttypes.php`

Frenziecodes commented:
> 

Modified the `'class-posttypes.php'` file to add the `'Album Creator Role'` setting under the `'Albums' tab` in the settings page. The role choices are fetched using the `'get_editable_roles'` function, and the default option is set to `'Inherit from gallery creator role'`. 

Updated the `'register_posttype'` function to use the `'capabilities'` parameter in the album post type registration. The capability is based on the selected `'Album Creator Role'` setting. If a specific role is chosen, it will be the only role capable of creating albums. If the setting is set to 'Inherit from gallery creator role', then it will act the same as the gallery and what that role is set to.

bradvin commented on extensions/albums/class-posttypes.php:
```diff
@@ -8,43 +8,84 @@
 			);
 
+			if ( 'inherit' !== $album_creator_role && '' !== $album_creator_role ) {
+				$args['capabilities'] = array(
```
Why are there more capabilities listed here for the album CPT, compared to the gallery CPT? For the gallery CPT, you used 
'create_posts' => $gallery_creator_role,
'create_post'  => $gallery_creator_role,
'edit_posts'   => $gallery_creator_role,
'edit_post'    => $gallery_creator_role,
'delete_post'  => $gallery_creator_role,
'delete_posts' => $gallery_creator_role,
'read-post' => $gallery_creator_role

bradvin commented on includes/admin/class-settings.php:
```diff
@@ -200,6 +200,29 @@ function create_settings() {
+				'choices' => $role_choices,
+				'default' => 'inherit',
+				'tab'     => 'album',
+				'section' => __( 'Admin', 'foogallery' ),
```
should not have a section under 'Albums' tab

bradvin commented on includes/admin/class-settings.php:
```diff
@@ -200,6 +200,29 @@ function create_settings() {
+				'type'    => 'select',
+				'choices' => $role_choices,
+				'default' => 'inherit',
+				'tab'     => 'album',
```
should be 'albums', otherwise there are 2 tabs for Albums:

![image](https://github.com/fooplugins/foogallery/assets/1409490/1e7a4b7b-cf62-4e13-8dd4-e368ea6df99d)
